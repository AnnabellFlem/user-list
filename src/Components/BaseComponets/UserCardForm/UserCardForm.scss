@import 'src/Styles/sizes';
@import 'src/Styles/colors';
@import 'src/Styles/media';

.user {
  &__btn {
    &--cancel {
      border: 1px solid $black-40;
      background-color: $black-5;
      color: $white-10;
    }

    &--create,
    &--save {
      border: 1px solid $black-5;
      background-color: $white-10;

      &:hover {
        background-color: darken($white-10, 5);
      }
    }

    &--cancel,
    &--save,
    &--create {
      border-radius: $border-radius-s;
      padding: $indent-xxs $indent-xs;
      color: $black-100;
      font-size: $font-xs;
      margin: $indent-s $indent-m;
      transition: all 0.2s ease;

      &:hover {
        box-shadow: 0 5px 10px $black-5;
      }
    }
  }

  &__form {

    &__info:first-of-type {
      padding-top: 0;
    }

    .user__data {
      margin: $indent-xs 0 $indent-xs $indent-s;
      padding: $indent-xxs $indent-xs;
      border-radius: $border-radius-s;
      font-size: $font-xs;
      border: 1px solid $black-40;
      text-align: start;
      display: inline-flex;
      width: 100%;
    }

    .user__title {
      display: inline-flex;
      align-items: center;

      @include tablet {
        width: 100%;
      }
    }

    .user__info {
      padding: $indent-xxs $indent-s;
    }

    &--main {
      .user__info {
        flex-direction: column;
        justify-content: start;
        align-items: start;
      }

      .user__data {
        display: inline-flex;
        margin: $indent-xs 0 $indent-xs $indent-l;

        @include tablet {
          display: flex;
          margin: $indent-xs 0 $indent-xs $indent-s;
        }
      }

      .user__title {
        display: flex;
        align-items: baseline;
      }
    }
  }

  &:nth-child(3n + 2) {
    margin-left: $indent-l * 2;
    margin-right: $indent-l * 2;

    @include desktop {
      margin-left: 0;
      margin-right: 0;
    }
  }

  &:nth-child(even) {
    @include desktop {
      margin-left: auto;
    }
  }
}
