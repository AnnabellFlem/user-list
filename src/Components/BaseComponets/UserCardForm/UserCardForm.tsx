import React from 'react'
import { useFormik } from 'formik'
import * as yup from 'yup'
import valid from 'card-validator'
import { User } from '../../../Types/User'
import { initialUserValues } from '../../../Utils/getInitialUserData'
import ErrorMessage from '../ErrorMessage'
import { usersRef } from '../../../Services/firebase'

import '../UserCard/UserCard.scss'
import '../UserCardForm/UserCardForm.scss'

const phoneRegExp = /^\+[(]\d{1,3}[)][ ]\d{3}-\d{3}-\d{2}-\d{2}$/
const UserSchema = yup.object().shape({
  name: yup.string().max(16).required(),
  surname: yup.string().max(20).required(),
  email: yup
    .string()
    .email('Please enter your valid email')
    .required('Please enter your email'),
  phone: yup
    .string()
    .matches(
      phoneRegExp,
      'Phone number is not valid, enter format +({country code}) XXX-XXX-XX-XX',
    ),
  credit: yup
    .string()
    .test(
      'test-number',
      'Credit Card number is invalid',
      value => valid.number(value).isValid,
    ),
})

type UserCardFormProps = {
  user?: User
  isMainEditForm?: boolean
  handleCreateClick?: any
  handleSaveClick?: any
  handleCancelClick?: any
}

const UserCardForm: React.FC<UserCardFormProps> = ({
  user,
  isMainEditForm,
  handleCreateClick,
  handleSaveClick,
  handleCancelClick,
}) => {
  const formik = useFormik({
    initialValues: user || initialUserValues,
    validationSchema: UserSchema,
    onSubmit: (values, { resetForm }) => {
      if (isMainEditForm) {
        usersRef.push(values)
      } else {
        usersRef
          .child(`${values.id}`)
          .set(values)
          .then(() => console.log('Data added successfully'))
          .catch(error => console.log('Adding failed with error' + error))
      }
      resetForm()
    },
  })

  return (
    <form
      className={`${isMainEditForm ? 'user__form--main' : ''} user__form`}
      onSubmit={formik.handleSubmit}
    >
      <div className="user__info">
        <label className="user__title">
          Name
          <input
            id="name"
            name="name"
            type="text"
            onChange={formik.handleChange}
            value={formik.values.name}
            className="user__data"
          />
        </label>
        {formik.errors.name && (
          <ErrorMessage>{formik.errors.name}</ErrorMessage>
        )}
      </div>
      <div className="user__info">
        <label className="user__title">
          Surname
          <input
            id="surname"
            name="surname"
            type="text"
            onChange={formik.handleChange}
            value={formik.values.surname}
            className="user__data"
          />
        </label>
        {formik.errors.surname && (
          <ErrorMessage>{formik.errors.surname}</ErrorMessage>
        )}
      </div>
      <div className="user__info">
        <label className="user__title">
          Email
          <input
            id="email"
            name="email"
            type="email"
            onChange={formik.handleChange}
            value={formik.values.email}
            className="user__data"
          />
        </label>
        {formik.errors.email && (
          <ErrorMessage>{formik.errors.email}</ErrorMessage>
        )}
      </div>
      <div className="user__info">
        <label className="user__title">
          Phone
          <input
            id="phone"
            name="phone"
            type="tel"
            onChange={formik.handleChange}
            value={formik.values.phone}
            className="user__data"
          />
        </label>
        {formik.errors.phone && (
          <ErrorMessage>{formik.errors.phone}</ErrorMessage>
        )}
      </div>
      <div className="user__info">
        <label className="user__title">
          Birthday
          <input
            id="birthday"
            name="birthday"
            type="date"
            onChange={formik.handleChange}
            value={formik.values.birthday}
            className="user__data"
            min="1950-01-01"
            max="2018-12-31"
          />
        </label>
      </div>
      <div className="user__info">
        <label className="user__title">
          Credit card
          <input
            id="credit"
            name="credit"
            type="text"
            onChange={formik.handleChange}
            value={formik.values.credit}
            className="user__data"
          />
        </label>
        {formik.errors.credit && (
          <ErrorMessage>{formik.errors.credit}</ErrorMessage>
        )}
      </div>
      {isMainEditForm ? (
        <button
          type="submit"
          className="user__btn--create"
          onClick={handleCreateClick}
        >
          Create
        </button>
      ) : (
        <div>
          <button
            type="submit"
            className="user__btn--save"
            onClick={handleSaveClick}
          >
            Save
          </button>
          <button
            type="button"
            className="user__btn--cancel"
            onClick={handleCancelClick}
          >
            Cancel
          </button>
        </div>
      )}
    </form>
  )
}

export default UserCardForm
