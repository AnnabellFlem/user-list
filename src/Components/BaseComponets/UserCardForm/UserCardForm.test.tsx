import React from 'react'
import { shallow } from 'enzyme'
import UserCardForm from './UserCardForm'

const mockedData = {
  id: 'id',
  name: 'name',
  surname: 'surname',
  email: 'email',
  phone: 'phone',
  birthday: 'birthday',
  credit: 'credit',
}

describe('UserCardForm test', () => {
  let wrapper: any

  beforeAll(() => {
    wrapper = shallow(<UserCardForm user={mockedData} />)
  })

  it('should to match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Component should render form', () => {
    expect(wrapper.find('form')).toHaveLength(1)
  })

  it('Component should render user info in input values', () => {
    expect(wrapper.find('input').at(0).props().value).toBe(mockedData.name)
    expect(wrapper.find('input').at(1).props().value).toBe(mockedData.surname)
  })
})
