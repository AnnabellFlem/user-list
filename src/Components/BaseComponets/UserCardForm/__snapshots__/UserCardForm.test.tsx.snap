// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCardForm test should to match snapshot 1`] = `ShallowWrapper {}`;
